pkg/server/bundle/pubmanager/pubmanager.go

```diff
@@ -0,0 +1,192 @@
+// Package pubmanager manages the publishing of the trust bundle to external
+// stores through the configured BundlePublisher plugins.
+package pubmanager
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/andres-erbsen/clock"
+	"github.com/sirupsen/logrus"
+	"github.com/spiffe/go-spiffe/v2/spiffeid"
+	"github.com/spiffe/spire/pkg/server/datastore"
+	"github.com/spiffe/spire/pkg/server/plugin/bundlepublisher"
+	"github.com/spiffe/spire/proto/spire/common"
+)
+
+const (
+	// refreshInterval is the interval to check for an updated trust bundle.
+	refreshInterval = 30 * time.Second
+)
+
+// NewManager creates a new bundle publishing manager.
+func NewManager(c *ManagerConfig) *Manager {
+	return newManager(c)
+}
+
+// ManagerConfig is the config for the bundle publishing manager.
+type ManagerConfig struct {
+	BundlePublishers []bundlepublisher.BundlePublisher
+	DataStore        datastore.DataStore
+	Clock            clock.Clock
+	Log              logrus.FieldLogger
+	TrustDomain      spiffeid.TrustDomain
+}
+
+// Manager is the manager for bundle publishing. It implements the PubManager
+// interface.
+type Manager struct {
+	bundleUpdatedCh  chan struct{}
+	bundlePublishers []bundlepublisher.BundlePublisher
+	clock            clock.Clock
+	dataStore        datastore.DataStore
+	log              logrus.FieldLogger
+	trustDomain      spiffeid.TrustDomain
+
+	hooks struct {
+		// Test hook used to indicate an attempt to publish a bundle using a
+		// specific bundle publisher.
+		publishResultCh chan *publishResult
+
+		// Test hook used to indicate when the action of publishing a bundle
+		// has finished.
+		publishedCh chan error
+	}
+}
+
+// Run runs the bundle publishing manager.
+func (m *Manager) Run(ctx context.Context) error {
+	if m.dataStore == nil {
+		return errors.New("missing datastore")
+	}
+
+	if m.trustDomain.IsZero() {
+		return errors.New("missing trust domain")
+	}
```

Comment:
These checks would probably make more sense now happening in NewManager since there is no longer an Init step.